### PR TITLE
Added 'Amazon Alexa Reviews Analysis' Project

### LINKED ISSUE #1028: Amazon Alexa Reviews Analysis
**Is your feature request related to a problem? Please describe.**
It's difficult and tedious for a company to analyze the huge amount of feedback that they get from their customer in terms of words (as reviews).

**Describe the solution you'd like**
In order to solve this issue for Amazon Alexa, I would like to add my project. Using NLP, the program can classify any feedback given to Alexa as a positive or a negative one and in turn can help the developers to automate the feedback process and improve on the service accordingly.

**Additional context**
My folder will be complete with the dataset, the notebook, model's pkl file and also a proper README.
The analysis looks something like this:
image

### PR DESCRIPTION
**Pull request by Manasi2001**
# Description

The folder structure goes like this:

Amazon Alexa Reviews Analysis

▫Dataset
- amazon_alexa.csv

▫Model
- Amazon_Alexa_Reviews_Analysis.py                 (NLP is performed, model built and accuracy checked)
- AmazonAlexaReviewsAnalysis.pkl                     (pickle file for storing the model)
- prediction.py                                                     (model is called, analysis of the review entered by user is performed)

▫README

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines(Clean Code) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have created a helpful and easy to understand `README.md`
- [x] My documentation follows [`Template for README.md`](https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/master/Template%20for%20README.md)

## Project Metadata

`` If there is no-file/nothing to fill the below fields with, then type: none ``

`` Example: `` If no requirements.txt needed/present, then type none in ``Requirments``.

Category:
- [ ] Calculators
- [x] AI/ML
- [ ] Scrappers
- [ ] Social_Media
- [ ] PDF
- [ ] Image_Processing
- [ ] Video_Processing
- [ ] Games
- [ ] Networking
- [ ] OS_Utilities
- [ ] Automation
- [ ] Cryptography
- [ ] Computer_Vision
- [ ] Fun
- [ ] Others

Title: Amazon Alexa Reviews Analysis

Folder: Model

Requirements: None

Script: prediction.py

Arguments: None

Contributor: Manasi2001

Description: The script has been made to perform NLP on Alexa's reviews. The user enters his/her review, and the script let him/her know whether the feedback was positive or negative.

### CODE REVIEW

File: Amazon Alexa Reviews Analysis/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+1. Removing stop words: A stop word is a commonly used word (such as “the”, “a”, “an”, “in”) that a search engine has been programmed to ignore, both when indexing entries for searching and when retrieving them as the result of a search query. We would not want these words to take up space in our database, or taking up valuable processing time. For this, we can remove them easily, by storing a list of words that you consider to stop words.
+
+2. Tokenization: Tokenization is essentially splitting a phrase, sentence, paragraph, or an entire text document into smaller units, such as individual words or terms. Each of these smaller units are called tokens.
+
+3. Stemming: Stemming is the process of reducing a word to its word stem that affixes to suffixes and prefixes or to the roots of words known as a lemma.
+
+## Source
+
+https://www.kaggle.com/sid321axn/amazon-alexa-reviews
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/README.md:25]:
```suggestion
[Amazon Alexa Reviews- Kaggle](https://www.kaggle.com/sid321axn/amazon-alexa-reviews)
```

File: Amazon Alexa Reviews Analysis/Model/Amazon_Alexa_Reviews_Analysis.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
+# Amazon Alexa Reviews Analysis
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/Model/Amazon_Alexa_Reviews_Analysis.py:1]: shift this file to the Amazon Alexa Reviews Analysis folder. Also, rename this file to training.py

File: Amazon Alexa Reviews Analysis/Model/Amazon_Alexa_Reviews_Analysis.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+from nltk.tokenize import TweetTokenizer   # module for tokenizing strings
+
+from sklearn.linear_model import LogisticRegression  
+from sklearn.feature_extraction.text import CountVectorizer  
+from sklearn.model_selection import train_test_split  
+from sklearn.metrics import roc_auc_score
+
+# Getting our Data
+
+df = pd.read_csv(r'C:\Users\DELL\Desktop\Kaggle+HE\Github GSSoC21\Amazing Python Scripts\Amazon Alexa Reviews Analysis\Dataset\amazon_alexa.csv')
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/Model/Amazon_Alexa_Reviews_Analysis.py:24]: use relative paths instead of your local system path. 
`./Amazon Alexa Reviews Analysis/dataset/amazon_alexa.csv`

File: Amazon Alexa Reviews Analysis/Model/Amazon_Alexa_Reviews_Analysis.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+# Checking Accuracy
+
+print('\nAccuracy:',roc_auc_score(y_test, y_pred))
+print('\n')
+
+# Predictions are 68.25% accurate.
+
+# Getting the pkl File
+import pickle
+pickle.dump(classifier, open('AmazonAlexaReviewsAnalysis.pkl', 'wb'))
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/Model/Amazon_Alexa_Reviews_Analysis.py:101]:
```suggestion
pickle.dump(classifier, open('./Amazon Alexa Reviews Analysis/model/AmazonAlexaReviewsAnalysis.pkl', 'wb'))
```

File: Amazon Alexa Reviews Analysis/Model/prediction.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import pandas as pd
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/Model/prediction.py:1]: Shift this file to Amazon Alexa Reviews Analysis folder

File: Amazon Alexa Reviews Analysis/Model/prediction.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import pandas as pd
+import Amazon_Alexa_Reviews_Analysis as A
+
+# loading our model
+
+import joblib
+classifier = joblib.load(r'C:\Users\DELL\Desktop\Kaggle+HE\Github GSSoC21\Amazing Python Scripts\Amazon Alexa Reviews Analysis\Model\AmazonAlexaReviewsAnalysis.pkl')
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/Model/prediction.py:7]: again, use relative paths as instructed in previous suggestion

File: Amazon Alexa Reviews Analysis/Model/prediction.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import pandas as pd
+import Amazon_Alexa_Reviews_Analysis as A
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/Model/prediction.py:2]: as the file will be renamed, update it here also. Also, if you use an alias for this then use a better word.

File: Amazon Alexa Reviews Analysis/Model/prediction.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+df1 = pd.read_csv(r'C:\Users\DELL\Desktop\Kaggle+HE\Github GSSoC21\Amazing Python Scripts\Amazon Alexa Reviews Analysis\Dataset\amazon_alexa.csv')
+
+l = []
+for i in df1['verified_reviews']:
+    l.append(i)
+    
+b = None    
+for j in range(0,len(l)):
+    if l[j] == a:
+        b = classifier.predict(A.X_train[j])
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/Model/prediction.py:21]: What is this logic? The user can give any input regardless it is present in the dataset or not. Here, you are only considering the dataset values. Change this to take new inputs from user

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]

File: Amazon Alexa Reviews Analysis/Model/prediction.py (hunk 8 above)
Comment (Manasi2001) [reply Amazon Alexa Reviews Analysis/Model/prediction.py:21]: I'm sorry. I spent like 3 to 4 hours yesterday trying to analyze a single review using the 'classifier', but it just wasn't working. I kept getting the same error no matter how many methods or manipulations I tried. This seemed to be the only final resolve that was working. 
I'll just try again. :(

Comment (kaustubhgupta) [reply Amazon Alexa Reviews Analysis/Model/prediction.py:21]: I think you need the same preprocessing pipelines which you used to train the dataset. Maybe the new text is not getting processed

Comment (Manasi2001) [reply Amazon Alexa Reviews Analysis/Model/prediction.py:21]: Did you check these new screenshots that I added? @kaustubhgupta Can something be done?

Comment (kaustubhgupta) [reply Amazon Alexa Reviews Analysis/Model/prediction.py:21]: The problem is here is that the model has been trained for 1500 features but the input provided is 2 features. This can be solved by using this:

(lines 72 to 75 in Amazon Alexa Reviews Analysis/Model/Amazon_Alexa_Reviews_Analysis.py)

cv = CountVectorizer(max_features=1500, analyzer='word', lowercase=False) 

df['verified_reviews'] = df['verified_reviews'].apply(lambda x: " ".join(x) )  # to join all words in the lists

X = cv.fit_transform(df['verified_reviews'])

You need to use count vectorizer transform function to transform input to 1500 features and then fed it to model. 

I would suggest that you create a single file for modeling and prediction.

**Comment by Manasi2001** [comment]
@kaustubhgupta it's finally working now! Thanks a bunch. Please check the updated project.

File: Amazon Alexa Reviews Analysis/README.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+3. Stemming: Stemming is the process of reducing a word to its word stem that affixes to suffixes and prefixes or to the roots of words known as a lemma.
+
+## Source
+
+[Amazon Alexa Reviews- Kaggle](https://www.kaggle.com/sid321axn/amazon-alexa-reviews)
+
+## Author
+
+Manasi Chhibber
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/README.md:29]:
```suggestion
[Manasi Chhibber](https://github.com/Manasi2001)

```

File: Amazon Alexa Reviews Analysis/reviews_analysis.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
+# Amazon Alexa Reviews Analysis
+
+# The aim is to analyse Alexa's reviews by NLP. If the feedback is positive, the result is 1, else it is 0. Using logistic regression, I have tried to classify the feedback as positive or negative.
+
+## MODELLING
+
+# Importing Libraries
+
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/reviews_analysis.py:8]: You can remove these comments

File: Amazon Alexa Reviews Analysis/reviews_analysis.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+df = pd.read_csv('../Amazon Alexa Reviews Analysis/Dataset/amazon_alexa.csv')
+
+# Data Preprocessing
+
+df = df.drop(['rating', 'date', 'variation'], axis = 1)
+
+df.isnull().any()  # checking for null values
+
+# no null values were found
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/reviews_analysis.py:34]: not needed

File: Amazon Alexa Reviews Analysis/reviews_analysis.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+            # rev_clean.append(word)
+            stem_word = stemmer.stem(word)  # stemming word
+            rev_clean.append(stem_word)
+
+    return rev_clean
+
+# using the process_rev function for:
+# 1. Removing stop words
+# 2. Tokenization
+# 3. Stemming
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/reviews_analysis.py:64]: shift this explanation in function doc string

File: Amazon Alexa Reviews Analysis/reviews_analysis.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+A = []
+a = df['verified_reviews']
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/reviews_analysis.py:66]: can we use better variable names here? The ones which explain the code easily?

File: Amazon Alexa Reviews Analysis/reviews_analysis.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+# Making Predictions
+
+y_pred = classifier.predict(X_test)
+
+# Checking Accuracy
+
+print('\nAccuracy:',roc_auc_score(y_test, y_pred))
+print('\n')
+
+# Predictions are 68.25% accurate.
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/reviews_analysis.py:99]: not needed for this script

File: Amazon Alexa Reviews Analysis/reviews_analysis.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+
+# Getting the pkl File
+import pickle
+pickle.dump(classifier, open('../Amazon Alexa Reviews Analysis/Model/AmazonAlexaReviewsAnalysis.pkl', 'wb'))
+
+## MAKING PREDICTIONS
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/reviews_analysis.py:105]: now that the model and prediction file is the same, you can skip the model pickling part. Also, remove the `/Model/AmazonAlexaReviewsAnalysis.pkl` file :)

File: Amazon Alexa Reviews Analysis/reviews_analysis.py [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+
+a = input('Enter review: ')
+
+# processing the review
+
+b = process_rev(a)
+
+b_new = ''
+for i in b:
+     b_new = b_new + ' ' + i
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/reviews_analysis.py:117]: again, better variable names

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]
Changes to be done:

**Review by kaustubhgupta — COMMENTED** [review]
Also, add the requirements.txt file

**Comment by Manasi2001** [comment]
Updated @kaustubhgupta.

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]
Umm,

image

**Review by kaustubhgupta — COMMENTED** [review]
You are using a highly imbalanced dataset:

image

Consider sampling or lookout for another one

File: Amazon Alexa Reviews Analysis/reviews_analysis.py [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+from nltk.tokenize import TweetTokenizer   # module for tokenizing strings
+
+from sklearn.linear_model import LogisticRegression  
+from sklearn.feature_extraction.text import CountVectorizer  
+from sklearn.model_selection import train_test_split  
+from sklearn.metrics import roc_auc_score
+
+# Getting our Data
+
+df = pd.read_csv('../Amazon Alexa Reviews Analysis/Dataset/amazon_alexa.csv')
```
Comment (kaustubhgupta) [inline Amazon Alexa Reviews Analysis/reviews_analysis.py:24]:
```suggestion
df = pd.read_csv('./Amazon Alexa Reviews Analysis/Dataset/amazon_alexa.csv')
```

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]

**Comment by kaustubhgupta** [comment]
@Manasi2001 Updates on this?

**Comment by Manasi2001** [comment]
> @Manasi2001 Updates on this?

I'm not willing to work on this anymore @kaustubhgupta.

**Comment by kaustubhgupta** [comment]
> > @Manasi2001 Updates on this?
> 
> I'm not willing to work on this anymore @kaustubhgupta.

ok
